### ISSUE TITLE
iterable.Sticky should be deleted

### ISSUE THREAD

**Issue by victornoel** [yegor256/cactoos#1334]
If one needs a "sticky" iterable, one can use `ListOf`, `SetOf` or any future collection implementation available, but have a `Sticky` iterable only creates confusion about the meaning of the `Iterable` interface and how it should be used.

I thus propose to remove this to simplify cactoos and avoid confusion.

**Comment by 0crat** [comment]
@paulodamaso[/z](https://www.0crat.com/u/paulodamaso) please, pay attention to this issue

**Comment by paulodamaso** [comment]
@victornoel `Sticky` implementations are designed to provide immutable objects. I think that  we should replace their usage with the  `Immutable`  implementatons, WDYT?

**Comment by victornoel** [comment]
@paulodamaso to me those are two perpendicular concerns: the objective of `iterable.Sticky` is to ensure an `Iterable` is iterated only once (and then next iterations would simply be based on the cached first one) while `Immutable` (which should IMHO be called `Unmodifiable`) is to forbid modification of the wrapped `Iterable` without consideration about stickiness.

If `Sticky` until now was responsible of providing also immutability, I think it was a mistake because one class should cover one responsibility. If needed Sticky and Immutable can be composed.

What I propose here is that `iterable.Sticky` should be removed and replaced with whatever makes sense, depending on the context. Be it `Immutable` (if it was used for immutability) or `ListOf` (if it was used for stickiness) or any other collection implementation.

**Comment by paulodamaso** [comment]
@victornoel You are right. Looking at `Sticky` docs in `org.cactoos.scalar.Scalar`: sticky are decorator used for caching. But I don't think that this should be the default behavior for iterables and lists; if we put this behavior into `ListOf`, we will end up by doing this. WDYT?

**Comment by victornoel** [comment]
@paulodamaso I'm not sure I understood the problem you are highlighting.

`ListOf` is by definition a stateful collection: if you give it an `Iterable`, it will copy its content inside the `List` represented by `ListOf`.

I am not proposing to modify `ListOf` in any way, I'm saying that `iterable.Sticky` is useless because its behaviour is already covered by `ListOf` or `SetOf`.

**Comment by paulodamaso** [comment]
@victornoel I understood now. I just think that is weird that we have `Sticky` behavior  automatically into `ListOf`. But this issue makes sense, if we already have the behaviorin `ListOf` there is no need for `Sticky`

**Comment by victornoel** [comment]
@paulodamaso actually, note that there is one small difference between `Sticky` here and using something like `ListOf`:
- with `ListOf`, the passed `Iterable` is copied during instantiation of `ListOf`
- with `Sticky`, the passed `Iterable` is copied only at the first traversal of the sticky iterable!
This could warrant keeping `Sticky` maybe...

**Comment by paulodamaso** [comment]
@victornoel Yes, seems that you are right. Don't you think that it could be interesting to have `ListOf` to behave like this too, for lazy evaluation's sake?

**Comment by victornoel** [comment]
@paulodamaso well it would go against everything we said before:
- that `Iterable`/`Iterator` are the actual OO abstractions we should on because they are lazy by nature
- that `ListOf`, `SetOf` and co are concrete collections and we should not for now put too much effort in there and try to implement a collection library until the rest of cactoos is more mature

So.. no :) and thus `Sticky` should be kept against my initial recommendation :)

**Comment by paulodamaso** [comment]
@victornoel Right, so let's close this one as is

**Comment by 0crat** [comment]
Job `gh:yegor256/cactoos#1334` is not assigned, can't get performer
